extension/src-language-server/fta/analysis/fta-cutSet-calculator.ts

```diff
@@ -0,0 +1,255 @@
+    // TODO: add minimal flag (could reduce computation cost)
+    const allCutSets = determineCutSetsForFT(allNodes);
+
+    // Cut sets are minimal if removing one element destroys the cut set
+    // If cut set contains another cut set from the array, remove it since it is not minimal
+    const minimalCutSet = allCutSets.filter((cutSet) => checkMinimalCutSet(cutSet, allCutSets));
+
+    return minimalCutSet;
+}
+
+/**
+ * Checks whether the given {@code cutSet} is a minimal cut set.
+ * @param cutSet The list to check.
+ * @param allCutSets All Cut Sets of the Fault Tree.
+ * @returns True if the given set is a minimal cut set.
+ */
+function checkMinimalCutSet(cutSet: Set<namedFtaElement>, allCutSets: Set<namedFtaElement>[]): boolean {
+    for (const otherCutSet of allCutSets) {
+        let contained = true;
+        otherCutSet.forEach((element) => {
+            if (!cutSet.has(element)) {
+                contained = false;
+            }
+        });
+        if (contained && cutSet !== otherCutSet) {
+            return false;
+        }
+    }
+    return true;
+}
+
+/**
+ * Determines all cut sets of a fault tree.
+ * @param allNodes All nodes of the fault tree.
+ * @returns the cut sets of the fault tree.
+ */
+export function determineCutSetsForFT(allNodes: AstNode[]): Set<namedFtaElement>[] {
+    /*  Idea:
+            Start from the top event.
+            Get the only child of top event (will always be only one) as our starting node.
+            Calculate all children of the node and evaluate them.
+            In the evaluation we check if the child has children too and do the same recursively until 
+            the children are components.
+            Depending on the type of the node process the results of the children differently. */
+
+    const startNode = getChildOfTopEvent(allNodes);
+    if (startNode) {
+        // determine the cut sets of the Fault Tree
+        return determineCutSetsForGate(startNode, allNodes);
+    }
+    return [];
+}
+
+/**
+ * Determines the cut sets for the (sub) fault tree that has {@code startNode} as the top node.
+ * @param startNode The top node of the (sub) fault tree for which the cut sets should be determined.
+ * @param allNodes All nodes of the fault tree.
+ * @param idCache The idCache of the generator context from the corresponding graph.
+ * @returns the determined cut sets for the (sub) fault tree as a list of lists.
+ */
+function determineCutSetsForGate(startNode: AstNode, allNodes: AstNode[]): Set<namedFtaElement>[] {
+    let result: Set<namedFtaElement>[] = [];
+
+    // components do not have children, so return the component
+    if (isComponent(startNode) || isCondition(startNode)) {
+        return [new Set<namedFtaElement>([startNode])];
+    }
+
+    const children = getChildrenOfNode(startNode);
+    if (children.length === 0 || !isGate(startNode)) {
+        return result;
+    }
+
+    if (isAND(startNode) || isInhibitGate(startNode)) {
```

Comment:
this if(isA)... else if(isB) etc.. structure could be replaced by a switch case to be more readable I think